### PR TITLE
[ll] Depth bounds test

### PR DESCRIPTION
**Pull request by jbatez**
Stubs and DirectX 12 amd Vulkan implementations of #1737

PR checklist:
- [x] `make` succeeds (on *nix)
- [x] `make reftests` succeeds
- [x] tested examples with the following backends: `dx12`, `vulkan`

### CODE REVIEW

**Comment by jbatez** [comment]
Went ahead and added the `dx12` implementation as well. As discussed in #1737, I had to add a bunch of code to the pipeline state creation just to set `DepthBoundsTestEnable: TRUE`.

**Comment by jbatez** [comment]
CreatePipelineState is returning E_INVALIDARG and I can't figure out why. :(

If anyone else wants to take a look it'd be much appreciated. :)

**Comment by kvark** [comment]
@jbatez you'll get a hint by checking DX debug runtime output. I know 2 ways to get that:
  1. launch the app from visual studio (no need to have a project for sources, just the executable is needed)
  2. have [DebugView](https://docs.microsoft.com/en-us/sysinternals/downloads/debugview) running side by side and showing the messages

File: src/backend/dx12/src/lib.rs [hunk 1]
```diff
@@ -752,6 +752,19 @@ impl hal::Instance for Instance {
                     mem::size_of::<d3d12::D3D12_FEATURE_DATA_ARCHITECTURE>() as _)
             });
 
+            let depth_bounds_test_supported = {
+                let mut features2: d3d12::D3D12_FEATURE_DATA_D3D12_OPTIONS2 = unsafe { mem::zeroed() };
+                if winerror::S_OK == unsafe {
```
Comment (kvark) [inline src/backend/dx12/src/lib.rs]: let's move the function all out of the condition for clarity, we'll then be able to combine the check into `error == winerror::S_OK && features2.DepthBoundsTestSupported != 0`. Although, it depends on what other `winerror` things are expected here.

File: src/hal/src/command/raw.rs [hunk 2]
```diff
@@ -292,6 +292,9 @@ pub trait RawCommandBuffer<B: Backend>: Clone + Any + Send + Sync {
     /// Set the blend constant values dynamically.
     fn set_blend_constants(&mut self, pso::ColorValue);
 
+    /// Set the depth bounds test values dynamically.
+    fn set_depth_bounds(&mut self, min: f32, max: f32);
```
Comment (kvark) [inline src/hal/src/command/raw.rs]: how about `Range<f32>`?

**Comment by jbatez** [comment]
@kvark Thanks for the DebugView tip! After hours of trial an error, I figured out the subobjects need to be aligned the same as pointers: a fact that's not documented anywhere as far as I can tell! I submitted MSDN documentation feedback 😛

File: src/backend/dx12/src/device.rs [hunk 3]
```diff
@@ -1496,11 +1496,104 @@ impl d::Device<B> for Device {
 
         // Create PSO
         let mut pipeline = ptr::null_mut();
-        let hr = unsafe {
-            self.raw.clone().CreateGraphicsPipelineState(
-                &pso_desc,
-                &d3d12::IID_ID3D12PipelineState,
-                &mut pipeline as *mut *mut _ as *mut *mut _)
+        let hr = if desc.depth_stencil.unwrap_or(unsafe { mem::zeroed() }).depth_bounds {
```
Comment (kvark) [inline src/backend/dx12/src/device.rs]: nit: maps/unwrap_or is most useful when you have linear code. Once you need to process both branches, a simple `match` looks better (to me):
```rust
let hr = match desc.depth_stencil {
   Some(ds) if ds.depth_bounds => {...}
   _ => {...}
};
```

File: src/backend/dx12/src/device.rs [hunk 4]
```diff
@@ -1496,11 +1496,104 @@ impl d::Device<B> for Device {
...
+            // The DepthBoundsTestEnable option isn't available in the original D3D12_GRAPHICS_PIPELINE_STATE_DESC struct.
+            // Instead, we must use the newer subobject stream method.
+            match self.raw.cast::<d3d12::ID3D12Device2>() {
+                Err(hr) => hr,
+                Ok(device2) => {
+                    #[repr(C)]
+                    struct Subobject<T> {
+                        subobject_align: [usize; 0], // Subobjects must have the same alignment as pointers.
```
Comment (kvark) [inline src/backend/dx12/src/device.rs]: great find! :clap:

File: src/backend/dx12/src/device.rs [hunk 5]
```diff
@@ -1496,11 +1496,104 @@ impl d::Device<B> for Device {
...
+                        ib_strip_cut_value: Subobject<d3d12::D3D12_INDEX_BUFFER_STRIP_CUT_VALUE>,
+                        primitive_topology: Subobject<d3d12::D3D12_PRIMITIVE_TOPOLOGY_TYPE>,
+                        render_target_formats: Subobject<d3d12::D3D12_RT_FORMAT_ARRAY>,
+                        depth_stencil_format: Subobject<dxgiformat::DXGI_FORMAT>,
+                        sample_desc: Subobject<dxgitype::DXGI_SAMPLE_DESC>,
+                        node_mask: Subobject<UINT>,
+                        cached_pso: Subobject<d3d12::D3D12_CACHED_PIPELINE_STATE>,
+                        flags: Subobject<d3d12::D3D12_PIPELINE_STATE_FLAGS>,
+                    }
+                    let mut pss_stream = SubobjectStream {
```
Comment (kvark) [inline src/backend/dx12/src/device.rs]: this function body became quite heavy, with a lot of meat indented. Let's move the bulk of logic into a helper `create_pipeline_stream` method returning `SubobjectStream` (which also means moving the types out of the function body)

File: src/backend/dx12/src/device.rs [hunk 6]
```diff
@@ -1512,14 +1605,19 @@ impl d::Device<B> for Device {
         if ds_destroy { destroy_shader(ds); }
 
         if winerror::SUCCEEDED(hr) {
+            let mut baked_states = desc.baked_states.clone();
+            if !desc.depth_stencil.unwrap_or(unsafe { mem::zeroed() }).depth_bounds {
```
Comment (kvark) [inline src/backend/dx12/src/device.rs]: similarly, let's avoid `unsafe` here, e.g. by doing `desc.depth_stencil.map(|ds| ds.depth_bounds).unwrap_or(false)`

File: src/backend/dx12/src/device.rs (hunk 3 above)
Comment (jbatez) [reply src/backend/dx12/src/device.rs]: I thought about it but the `None` case is the same as the `false` case and I wasn't sure howto combine the two.

File: src/backend/dx12/src/device.rs (hunk 6 above)
Comment (jbatez) [reply src/backend/dx12/src/device.rs]: Thanks! Still newish to Rust and didn't know about `Option::map`.

**Comment by kvark** [comment]
Looks like all the concerns are addressed, thank you!
Let's make CI happy, remove the WIP part, and proceed.

**Comment by jbatez** [comment]
@kvark Still waiting on new `ash` release.

**Comment by kvark** [comment]
@MaikKlein can we have a new version published please?

**Comment by MaikKlein** [comment]
@kvark https://crates.io/crates/ash/0.24.0

**Comment by kvark** [comment]
Could you please merge the first and the last commit? That would make it so each commit is buildable on CI.

**Comment by jbatez** [comment]
Commits squashed!

**Comment by kvark** [comment]
Bors r+

**Pull request merged**